## Практическое домашнее задание на проверку знаний по модулю 1.


##### Тема тестирования - базовый эксплоративный анализ и аналитическое мышление. Вам будет предоставлен датасет и 25 вопросов касаемо данного датасета. Ваша задача - подумать как найти ответы на поставленные вопросы ниже. За каждый правильно оформленный и предоставленный ответ начисляется *4 балла*.

#### Ответ на решение присылать ссылкой на ваш гитхаб репозиторий на почту maratmovlamov2017@gmail.com

#### Сроки выполнения практического домашнего задания - ровно 7 календарных дней (168 часов) с момента публикации в группе. Прошу приступать к решению задач порционно. Прошу повторить дополнительные материалы, предоставленные в качестве самостоятельного изучения в конце презентаций, они вам понадобятся.



#### Критерии оценки присланного решения:

* Правильность полученного ответа - на каждый поставленный вопрос есть один и только один правильный ответ. В случае неправильного ответа на поставленный вопрос, предоставленный ответ на него считается проваленным, даже если код написан правильно и красиво. Общее количество баллов, которые можно получить ответив на все вопросы верно - 25 * 4 = 100 баллов, которые шкалируются к 10-ти бальной шкале.

___

* Аккуратность, правильность и читаемость кода - старайтесь писать код без повторного использования методов и ненужных переменных (помните про принцип DRY - don't repeat yourself). Плюсом будет, если ваш код будет упакован в лаконичную и красиво написанную функцию. Если вся работа выполненна аккуратно, без спагетти кода, к общему результату работы начисляется 1 балл.

---

* На предоставленные вопросы, дописывается комментарий для проверки вашего мышления и поиска ответа в виде Markdown комментария. В случае если комментарий на поставленный вопрос не предоставлен снимается 0.1 балл за каждый такой вопрос. То есть общий штраф за отсутсвие комментариев по всем вопросам 25 * -0.1 =  -2.5 балла.


#### Удачи и да прибудет с Вами сила :)



In [578]:
import pandas as pd
import numpy as np

In [579]:
df = pd.read_csv("./dataset/Spotify_Youtube.csv")

In [580]:
df.head()

Unnamed: 0    Artist                                        Url_spotify  \
0           0  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
1           1  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
2           2  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
3           3  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
4           4  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   

                                           Track  \
0                                 Feel Good Inc.   
1                                Rhinestone Eyes   
2  New Gold (feat. Tame Impala and Bootie Brown)   
3                             On Melancholy Hill   
4                                 Clint Eastwood   

                                           Album Album_type  \
0                                     Demon Days      album   
1                                  Plastic Beach      album   
2  New Gold (feat. Tame Impala and Bootie Brown)     single   
3                                  Plastic Beach      album   
4                                       Gorillaz      album   

                                    Uri  Danceability  Energy   Key  ...  \
0  spotify:track:0d28khcov6AiegSCpG5TuT         0.818   0.705   6.0  ...   
1  spotify:track:1foMv2HQwfQ2vntFf9HFeG         0.676   0.703   8.0  ...   
2  spotify:track:64dLd6rVqDLtkXFYrEUHIU         0.695   0.923   1.0  ...   
3  spotify:track:0q6LuUqGLUiCPP1cbdwFs3         0.689   0.739   2.0  ...   
4  spotify:track:7yMiX7n9SBvadzox8T5jzT         0.663   0.694  10.0  ...   

                                   Url_youtube  \
0  https://www.youtube.com/watch?v=HyHNuVaZJ-k   
1  https://www.youtube.com/watch?v=yYDmaexVHic   
2  https://www.youtube.com/watch?v=qJa-VFwPpYA   
3  https://www.youtube.com/watch?v=04mfKJWDSzI   
4  https://www.youtube.com/watch?v=1V_xRb0x9aw   

                                               Title   Channel        Views  \
0         Gorillaz - Feel Good Inc. (Official Video)  Gorillaz  693555221.0   
1  Gorillaz - Rhinestone Eyes [Storyboard Film] (...  Gorillaz   72011645.0   
2  Gorillaz - New Gold ft. Tame Impala & Bootie B...  Gorillaz    8435055.0   
3     Gorillaz - On Melancholy Hill (Official Video)  Gorillaz  211754952.0   
4         Gorillaz - Clint Eastwood (Official Video)  Gorillaz  618480958.0   

       Likes  Comments                                        Description  \
0  6220896.0  169907.0  Official HD Video for Gorillaz' fantastic trac...   
1  1079128.0   31003.0  The official video for Gorillaz - Rhinestone E...   
2   282142.0    7399.0  Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3  1788577.0   55229.0  Follow Gorillaz online:\nhttp://gorillaz.com \...   
4  6197318.0  155930.0  The official music video for Gorillaz - Clint ...   

   Licensed official_video        Stream  
0      True           True  1.040235e+09  
1      True           True  3.100837e+08  
2      True           True  6.306347e+07  
3      True           True  4.346636e+08  
4      True           True  6.172597e+08  

[5 rows x 28 columns]

In [581]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20718 entries, 0 to 20717
Data columns (total 28 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        20718 non-null  int64  
 1   Artist            20718 non-null  object 
 2   Url_spotify       20718 non-null  object 
 3   Track             20718 non-null  object 
 4   Album             20718 non-null  object 
 5   Album_type        20718 non-null  object 
 6   Uri               20718 non-null  object 
 7   Danceability      20716 non-null  float64
 8   Energy            20716 non-null  float64
 9   Key               20716 non-null  float64
 10  Loudness          20716 non-null  float64
 11  Speechiness       20716 non-null  float64
 12  Acousticness      20716 non-null  float64
 13  Instrumentalness  20716 non-null  float64
 14  Liveness          20716 non-null  float64
 15  Valence           20716 non-null  float64
 16  Tempo             20716 non-null  float6

In [582]:
unwanted_columns = [ 'Url_spotify', 'Uri', 'Url_youtube', 'Title', 'Description', 'official_video', 'Licensed', 'Acousticness', 'Instrumentalness', 'Liveness']
df.drop(unwanted_columns, axis=1, inplace=True)

In [583]:
median_loud, median_energy = df['Loudness'].median(), df['Energy'].median()
df['Loudness'].fillna(median_loud, inplace = True )
df['Energy'].fillna(median_energy, inplace = True )

In [584]:
df.isnull().sum()

Unnamed: 0        0
Artist            0
Track             0
Album             0
Album_type        0
Danceability      2
Energy            0
Key               2
Loudness          0
Speechiness       2
Valence           2
Tempo             2
Duration_ms       2
Channel         470
Views           470
Likes           541
Comments        569
Stream          576
dtype: int64

In [585]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20718 entries, 0 to 20717
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    20718 non-null  int64  
 1   Artist        20718 non-null  object 
 2   Track         20718 non-null  object 
 3   Album         20718 non-null  object 
 4   Album_type    20718 non-null  object 
 5   Danceability  20716 non-null  float64
 6   Energy        20718 non-null  float64
 7   Key           20716 non-null  float64
 8   Loudness      20718 non-null  float64
 9   Speechiness   20716 non-null  float64
 10  Valence       20716 non-null  float64
 11  Tempo         20716 non-null  float64
 12  Duration_ms   20716 non-null  float64
 13  Channel       20248 non-null  object 
 14  Views         20248 non-null  float64
 15  Likes         20177 non-null  float64
 16  Comments      20149 non-null  float64
 17  Stream        20142 non-null  float64
dtypes: float64(12), int64(1), 

### Структура датасета:

----

26 признаков, описание которых следует ниже:



- **Track**: name of the song, as visible on the Spotify platform.
- **Artist**: name of the artist.
- **Url_spotify**: the Url of the artist.
- **Album**: the album in wich the song is contained on Spotify.
- **Album_type**: indicates if the song is relesead on Spotify as a single or contained in an album.
- **Uri**: a spotify link used to find the song through the API.
- **Danceability**: describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.
- **Energy:** is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, death metal has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.
- **Key**: the key the track is in. Integers map to pitches using standard Pitch Class notation. E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.
- **Loudness**: the overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typically range between -60 and 0 db.
- **Speechiness**: detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.
- **Acousticness**: a confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.
- **Instrumentalness**: predicts whether a track contains no vocals. "Ooh" and "aah" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly "vocal". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.
- **Liveness**: detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.
- **Valence**: a measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).
- **Tempo**: the overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.
- **Duration_ms**: the duration of the track in milliseconds.
- **Stream**: number of streams of the song on Spotify.
- **Url_youtube**: url of the video linked to the song on Youtube, if it have any.
- **Title**: title of the videoclip on youtube.
- **Channel**: name of the channel that have published the video.
- **Views**: number of views.
- **Likes**: number of likes.
- **Comments**: number of comments.
- **Description**: description of the video on Youtube.
- **Licensed**: Indicates whether the video represents licensed content, which means that the content was uploaded to a channel linked to a YouTube content partner and then claimed by that partner. official_video: boolean value that indicates if the video found is the official video of the song.

### Вопросы:

#### 1. Какая песня является наиболее транслируемой в датасете (признак Stream)?

В методе loc указываем индекс и наименование столбца из которого нужно вывести значение
Метод idxmax() возвращает индекс максимального значения в колонке Stream

In [520]:
print(f'Наиболее транслируемая песня', df.loc[df['Stream'].idxmax(), 'Track']) 

Наиболее транслируемая песня Blinding Lights


#### 2. У какого исполнителя больше всего просмотров на YouTube?

Группируем датафрейм по исполнителям и суммируем количество просмотров (Views) с помощью метода idxmax() определяем индекс исполнителя, у которого максимальноe количество просмотров

In [521]:
max_view_artist = df.groupby('Artist').aggregate({'Views': 'sum'}).idxmax()
print('Больше всего просмотров на YouTube у', max_view_artist[0])

Больше всего просмотров на YouTube у Ed Sheeran


#### 3. Какова средняя продолжительность всех песен в минутах?

Находим среднее значение столбца Duration_ms, используя метод mean(), полученное значение делим на 60000 для перевода мс в мин

In [522]:
print(f'Средняя продолжительность всех песен', round(df['Duration_ms'].mean()/60000, 1), 'мин')

Средняя продолжительность всех песен 3.7 мин


#### 4. Какая нота (Key) чаще всего используется в песнях?

C помощью метода value_counts() считаем, сколько раз каждая нота встречается в столбце

In [523]:
value_key = df['Key'].value_counts()

Составляем словарь для соответствия каждого ключа Key ноте

In [524]:
note_dict = {0.0: 'C', 1.0: 'C♯, D♭', 2.0: 'D', 3.0: 'D♯, E♭', 4.0: 'E', 5.0: 'F', 6.0: 'F♯, G♭', 7.0: 'G', 8.0: 'G♯, A♭', 9.0: 'A', 10.0: 'A♯, B♭', 11.0: 'B'}

In [525]:
print('Чаще всего используется нота', note_dict[value_key.index[0]])

Чаще всего используется нота C


#### 5. Какая самая популярная песня на YouTube?

У нас есть данные по просмотрам, лайкам и комментариям на ютубе, все эти параметры будут вносить свой вклад в популярность песни. Если приводить к одному параметру, то непонятно, какой вклад будет вносить каждый из параметров.
Поэтому рассчитаем самые популярные песни на YouTube по количеству просмотров, комментариев и количеству лайков.

In [526]:
print(f'Самая популярная песня на YouTube по просмотрам', df.loc[df['Likes'].idxmax(), 'Track']) 
print(f'Самая популярная песня на YouTube по лайкам', df.loc[df['Views'].idxmax(), 'Track']) 
print(f'Самая популярная песня на YouTube по комментариям', df.loc[df['Comments'].idxmax(), 'Track']) 

Самая популярная песня на YouTube по просмотрам Despacito
Самая популярная песня на YouTube по лайкам Despacito
Самая популярная песня на YouTube по комментариям Dynamite


#### 6. У какой песни самый высокий балл за танцевальность (danceability )?

Определяем самымй большой показатель Danceability и по индексу выводим название песни

In [527]:
print(f'Cамый высокий балл за танцевальность у песни', df.loc[df['Danceability'].idxmax(), 'Track'])

Cамый высокий балл за танцевальность у песни Give It To Me


#### 7. У какой песни самый высокий показатель энергии (energy score)?

Находим максимальное значение Energy и по индексу находим наименование трека

In [528]:
print(f'У песни', df.loc[df['Energy'].idxmax(), 'Track'], 'самый большой показатель энергии')

У песни Rain and Thunderstorm, Pt. 3 самый большой показатель энергии


#### 8. Какая песня имеет наивысший балл valence (valence score)?

Находим максимальный показатель Valence и по индексу определяем название песни

In [529]:
print(f'Наивысший балл valence имеет песня', df.loc[df['Valence'].idxmax(), 'Track'])

Наивысший балл valence имеет песня Spider Dance


#### 9. Будут ли песни с более высокой «valence» (музыкальной позитивностью) более популярны как на Spotify, так и на YouTube? Чтобы ответить на этот вопрос, подсчитайте среднюю оценку «valence» для песен в верхнем квартиле «Streams» и «Views» и сравните ее с общей средней «valence».

Создадим два датафрейма, которые будут содержать данные valence верхних квартилей Views и Stream

In [530]:
views_25 = df[['Valence', 'Views']].sort_values('Views', ascending=False)[:5062]
stream_25 = df[['Valence', 'Stream']].sort_values('Stream', ascending=False)[:5036]

Найдем средние значения valence

In [531]:
print(f'Среднее значение valence равно', round(df['Valence'].mean(), 3))
print(f'Среднее значение valence в верхнем квартиле по просмотрам на Youtube равно', round(views_25['Valence'].mean(), 3))
print(f'Среднее значение valence в верхнем квартиле по стримам на Spotify равно', round(stream_25['Valence'].mean(), 3))

Среднее значение valence равно 0.53
Среднее значение valence в верхнем квартиле по просмотрам на Youtube равно 0.548
Среднее значение valence в верхнем квартиле по стримам на Spotify равно 0.524


Если песни с более высокой valence более популярны, то это означает, что должна быть положительная корреляция между valence и views, valence и stream. Найдем коэффицент корреляции для этих данных

In [532]:
cor_val_v = round(views_25['Valence'].corr(df['Views']), 3)
cor_val_s = round(stream_25['Valence'].corr(df['Stream']), 3)
print(f'Коэффициент корреляции между valence и views равен', cor_val_v)
print(f'Коэффициент корреляции между valence и stream равен', cor_val_s)

Коэффициент корреляции между valence и views равен 0.026
Коэффициент корреляции между valence и stream равен -0.016


Коэффициенты корреляции стремятся к нулю, следовательно линейной зависимости между этими параметрами, следовательно нельзя утверждать, что песни с более высокой valence будут более популярны.

#### 10. Какой канал на YouTube имеет наибольшее количество просмотров?

Группируем каналы и суммируем просмотры для каждого канала, находим макс значение по просмотрам и определяем канал

In [533]:
max_views_channel = df.groupby('Channel').aggregate({'Views': 'sum'}).idxmax()
print('Наибольшее количество просмотров на YouTube имеет канал', max_views_channel[0])

Наибольшее количество просмотров на YouTube имеет канал T-Series


#### 11. У песни какого исполнителя больше всего комментариев на YouTube?

Группируем исполнителей и суммируем количество комментариев для каждого исполнителя, находим максимальное значение просмотров на ютуб и определяем исполнителя

In [534]:
max_comments_artist = df.groupby('Artist').aggregate({'Comments': 'sum'}).idxmax()
print(f'Больше всего комментариев на YouTube у исполнителя', max_comments_artist[0])

Больше всего комментариев на YouTube у исполнителя BTS


#### 12. Какой самый распространенный темп (tempo) песен?

С помощью метода value_counts() определяем, сколько раз значения темпа встречаются в столбце

In [586]:
value_counts_tempo = df['Tempo'].value_counts()

In [587]:
print('Самый распространенный темп песен равен', value_counts_tempo.index[0])

Самый распространенный темп песен равен 77.986


#### 13. Какая песня имеет самый высокий показатель по speechiness?

С помощью метода idxmax() находим индекс максимального значения speechiness и определяем наименование трека

In [537]:
max_speechiness = df.loc[df['Speechiness'].idxmax(), 'Track']

In [538]:
print(f'Самый высокий показатель по speechiness имеет песня', df.loc[df['Speechiness'].idxmax(), 'Track'])

Самый высокий показатель по speechiness имеет песня Teil 7 - Sherlock Holmes und der blinde Bettler - Die neuen Abenteuer, Folge 50


#### 14. Какая самая громкая песня в наборе данных?

Громкость определяется по параметру Loudness. Чем выше этот показатель, тем громче песня. Находим максимальное значение показателя и по его индексу определяем песню.

In [539]:
print(f'Cамая громкая песня в наборе данных', df.loc[df['Loudness'].idxmax(), 'Track'])

Cамая громкая песня в наборе данных Siren


#### 15. У какого артиста самый высокий средний балл danceability?

Группируем данные по колонке исполнителя, при этом значения Danceability уседняем для каждого исполнителя, находим максимальное среднее значение и по нему определяем исполнителя

In [540]:
max_danc_artist = df.groupby('Artist').aggregate({'Danceability': 'mean'}).idxmax()
print('Самый высокий средний балл danceability у', max_danc_artist[0])

Самый высокий средний балл danceability у Saweetie


#### 16. Как средний показатель «speechiness» песен зависит от «Album_type»? Песни, выпущенные как синглы, чаще содержат больше слов согласно анализу?

Группируем по Album_type и определяем среднее значение Speechiness для каждого типа альбома

In [588]:
df.groupby('Album_type').aggregate({'Speechiness': 'mean'})

Speechiness
Album_type              
album           0.094900
compilation     0.078867
single          0.103867

Наименьший показатель Speechiness у альбомов-сборников (0.079), средний показатель у альбомов (0.095) и наибольший у синглов (0.104). Песни, выпущенные как синглы, чаще содержат больше слов, чем альбомы и альбомы-сборники.

#### 17. Каково среднее количество прослушиваний песен, которые входят в топ 10% лучших по метрике танцевальности?

Создадим новый датафрейм, состоящий из 2 колонок: Stream и упрорядоченного по убыванию Danceability и сделаем срез 10% значений датафрейма

In [594]:
df_dance10 = df[['Stream', 'Danceability']].sort_values('Danceability', ascending=False)[:2072]

Найдем среднее значение Stream  в полученном датафрейме 

In [590]:
print(f'Среднее количество прослушиваний песен, которые входят в топ 10% лучших по метрике танцевальности, равно', round(df_dance10['Stream'].mean()))

Среднее количество прослушиваний песен, которые входят в топ 10% лучших по метрике танцевальности, равно 151903593


#### 18. Для песен, которые входят в топ 10% лучших с точки зрения «danceability» метрики, как соотносится их средний показатель «tempo» со средним показателем «tempo» всех песен?

In [544]:
df_dance_10 = df[['Tempo', 'Danceability']].sort_values('Danceability', ascending=False)[:2072]

In [545]:
mean_tempo = round(df['Tempo'].mean(), 1)
mean_tempo_10 = round(df_dance_10['Tempo'].mean(), 1)
print ('Cредний показатель «tempo» равен', mean_tempo)
print ('Cредний показатель «tempo» для песен, которые входят в топ 10% лучших с точки зрения «danceability»  равен', mean_tempo_10)

Cредний показатель «tempo» равен 120.6
Cредний показатель «tempo» для песен, которые входят в топ 10% лучших с точки зрения «danceability»  равен 115.1


In [546]:
print(f'Коэффициент корреляции для темпа и танцевальности песен равен', round(df['Tempo'].corr(df['Danceability']), 3))
print(f'Коэффициент корреляции для темпа и топ 10% танцевальности песен равен', round(df_dance_10['Tempo'].corr(df['Danceability']), 3))

Коэффициент корреляции для темпа и танцевальности песен равен -0.066
Коэффициент корреляции для темпа и топ 10% танцевальности песен равен 0.035


В виду того, что полученные коэффициенты корреляции близки к нулю, можем сделать вывод, что темп и танцевальность не являются линейно зависимыми. Исходя, из средних значений темпа, можем сделать вывод, что песни с высокой танцевальностью имеют темп ниже, чем весь набор песен.

#### 19. Какова корреляция между «valence» метрикой песни и ее популярностью на YouTube (по показателю «Views») и Spotify (по показателю «Stream»)? Счастливые песни более популярны? Valency отвечает за жизнерадостность песни.

Найдем коэффициенты корреляции между Valence и Views и Valence и Stream

In [547]:
coef_corr_Val_V =  df['Valence'].corr(df['Views'])
coef_corr_Val_S = df['Valence'].corr(df['Stream'])
print(f'Коэффициент корреляции между Valence и Views равен', round(coef_corr_Val_V, 3))
print(f'Коэффициент корреляции между Valence и Stream равен', round(coef_corr_Val_S, 3))

Коэффициент корреляции между Valence и Views равен 0.035
Коэффициент корреляции между Valence и Stream равен -0.012


 Коэффициенты корреляции близки к нулю, это означает, что нет линейной зависимости между valency и популярностью песни.

#### 20. Какова средняя громкость песен, которые входят в топ 10% лучших по энергетике (Energy)?

Создадим датафрейм, состоящий из 2 колонок: Loudness и Energy, отсортированного по убыванию. Сделаем срез на 2072 значения - 10% от 20716

In [548]:
df3 = df[['Loudness', 'Energy']].sort_values('Energy', ascending=False)[:2072]

Найдем среднее значение громкости для топ 10% по энергетике

In [549]:
print(f'Cредняя громкость песен, которые входят в топ 10% лучших по энергетике, составляет', round(df3['Loudness'].mean(), 1), 'db')

Cредняя громкость песен, которые входят в топ 10% лучших по энергетике, составляет -4.6 db


#### 21. У какого исполнителя больше всего песен, которые входят в топ-10% как по просмотрам на YouTube, так и по стримам Spotify?

Найдем топ-10% Youtube по просмотрам и Spotify по стримам

In [550]:
top_10_youtube = df[['Unnamed: 0','Artist', 'Track', 'Views', 'Stream']].sort_values('Views', ascending=False)[:2025]
top_10_spotify = df[['Unnamed: 0', 'Artist', 'Track', 'Views', 'Stream']].sort_values('Stream', ascending=False)[:2014]

Далее соединим эти датафреймы в один и отбросим дубликаты

In [551]:
top_10 = pd.concat( [top_10_spotify,top_10_youtube])

In [552]:
top_10 = top_10.drop_duplicates(subset=['Unnamed: 0', 'Unnamed: 0'])

Полученный датафрейм сгруппируем по исполнителям и посчитаем количество строк для каждого

In [553]:
top_10.groupby('Artist').size().sort_values(ascending=False).head(13)

Artist
Coldplay                   10
Queen                      10
Linkin Park                10
Ariana Grande              10
Post Malone                10
Dua Lipa                   10
Earth, Wind & Fire         10
Maroon 5                   10
Avicii                     10
XXXTENTACION               10
Eminem                     10
Macklemore & Ryan Lewis    10
J Balvin                   10
dtype: int64

В результате мы получили 12 исполнителей, песни которых входят в топ-10% по просмотрам на Youtube и по стримам на Spotify: The Mamas & The Papas, Dua Lipa, Coldplay, Linkin Park, J Balvin, Eminem, Earth, Wind & Fire, Maroon 5, Ariana Grande, Macklemore & Ryan Lewis, Post Malone, Avicii, Khalid  

#### 22. Какова корреляция между «energy» песни и ее «loudness»?

Найдем коэффициент корреляции, используя метод .corr()

In [569]:
cor_E_L = df['Energy'].corr(df['Loudness'])

In [570]:
cor_E_L

0.7448445555528453

Коэффициент корреляции равен 0,745, что говорит о положительной корреляции.
Теперь нужно вычислить р-значение, чтобы понять является ли корреляции статистически значимой

In [573]:
from scipy.stats import pearsonr
pearsonr(df['Energy'], df['Loudness'])

PearsonRResult(statistic=0.7448450518541081, pvalue=0.0)

Так как pvalue=0.0, то можем сделать вывод о том, что существует значимая связь между переменными Energy и Loudness

#### 23. Каково среднее количество лайков на YouTube для песен, которые находятся в верхнем квартиле (25%) и как это соотносится с общим средним показателем?

In [557]:
likes_25 = df['Likes'].sort_values(ascending=False)[:5044]

In [558]:
print(f'Среднее значение лайков равно', round(df['Likes'].mean()))
print(f'Cреднее значение лайков в верхнем квартиле', round(likes_25.mean()))

Среднее значение лайков равно 663341
Cреднее значение лайков в верхнем квартиле 2307672


Среднее значение лайков в верхнем квартиле в 3,33 раза превышает количество лайков для всех песен

#### 24. Есть ли существенная разница в средней «valence» метрикой для песен, выпущенных как синглы, по сравнению с песнями, выпущенными в виде альбома?

Отфильтруем наши данные по типу альбома single и album

In [559]:
valence_single = df.query('Album_type == "single"')
valence_album = df.query('Album_type == "album"')

Найдем среднее значение Valence для синглов и альбомов

In [560]:
mean_val_single = round(valence_single['Valence'].mean(), 4)
mean_val_album = round(valence_album['Valence'].mean(), 4)
print(f'Среднее значение Valence для песен, выпущенных как синглы, равно', mean_val_single)
print(f'Среднее значение Valence для песен, выпущенных как альбомы, равно', mean_val_album)

Среднее значение Valence для песен, выпущенных как синглы, равно 0.5301
Среднее значение Valence для песен, выпущенных как альбомы, равно 0.5298


Полученные значения очень близки, рассчитаем t-критерий Стьюдента и p-value, чтоб определить, являются ли различия статистически значимыми.

In [561]:
from scipy.stats import ttest_ind
result = ttest_ind(valence_single['Valence'], valence_album['Valence'])
print(f'T-критерий Стьюдента равен', round(result[0], 3), ', p-value равен', round(result[1],3))

T-критерий Стьюдента равен nan , p-value равен nan


Значение p-value получилось равным 0,579, что превышает значение 0,05 и говорит о том, что нет статистически значимых различий между выборками

#### 25. Каково среднее количество «комментариев» на YouTube для песен с показателем «speechiness» выше 0,66 (что указывает на большое количество произнесенных слов) и как это соотносится с общим средним показателем?

 Отсортируем датафрейм по показателю Speechiness > 0.66

In [562]:
speech_066 = df.query('Speechiness > 0.66')

In [563]:
speech_066

Unnamed: 0                  Artist  \
469           469            Die drei ???   
470           470            Die drei ???   
471           471            Die drei ???   
472           472            Die drei ???   
474           474            Die drei ???   
...           ...                     ...   
15725       15725                   Ernia   
15773       15773         Leslie Odom Jr.   
16582       16582                Blueface   
18029       18029  A Boogie Wit da Hoodie   
19968       19968                  j-hope   

                                          Track  \
469    Teil 23 - Das Geheimnis der Geisterinsel   
470    Teil 10 - Das Geheimnis der Geisterinsel   
471     Teil 1 - Das Geheimnis der Geisterinsel   
472     Teil 4 - Das Geheimnis der Geisterinsel   
474     Teil 3 - Das Geheimnis der Geisterinsel   
...                                         ...   
15725                                     Bella   
15773                           Aaron Burr, Sir   
16582                                    TikTok   
18029                                    Jungle   
19968                                     Intro   

                                             Album Album_type  Danceability  \
469                 Das Geheimnis der Geisterinsel      album         0.518   
470                 Das Geheimnis der Geisterinsel      album         0.415   
471                 Das Geheimnis der Geisterinsel      album         0.509   
472                 Das Geheimnis der Geisterinsel      album         0.572   
474                 Das Geheimnis der Geisterinsel      album         0.584   
...                                            ...        ...           ...   
15725                    Come Uccidere Un Usignolo      album         0.561   
15773  Hamilton (Original Broadway Cast Recording)      album         0.646   
16582                                       TikTok     single         0.752   
18029                                       Artist      album         0.538   
19968                              Jack In The Box      album         0.607   

       Energy  Key  Loudness  Speechiness  ...  Instrumentalness  Liveness  \
469     0.930  1.0    -6.246        0.858  ...          0.000000    0.8840   
470     0.743  1.0    -9.428        0.836  ...          0.000000    0.3390   
471     0.744  7.0    -7.231        0.899  ...          0.000000    0.1560   
472     0.769  1.0    -9.557        0.761  ...          0.000000    0.6090   
474     0.482  1.0   -10.423        0.865  ...          0.000000    0.6640   
...       ...  ...       ...          ...  ...               ...       ...   
15725   0.577  6.0   -11.899        0.935  ...          0.000000    0.2000   
15773   0.439  7.0   -10.837        0.818  ...          0.000000    0.0945   
16582   0.561  3.0    -8.754        0.885  ...          0.000000    0.2200   
18029   0.520  7.0   -11.063        0.737  ...          0.000000    0.2140   
19968   0.355  5.0   -14.867        0.828  ...          0.000281    0.7060   

       Valence    Tempo  Duration_ms                 Channel       Views  \
469      0.241   63.653     180008.0       Serien Schätze de     21912.0   
470      0.376  175.024     180008.0       Serien Schätze de     21912.0   
471      0.540   50.977     185263.0       Serien Schätze de     21912.0   
472      0.453  110.304     180006.0       Serien Schätze de     21912.0   
474      0.638  103.357     180006.0       Serien Schätze de     21912.0   
...        ...      ...          ...                     ...         ...   
15725    0.646  172.307     206510.0                   Ernia   8883233.0   
15773    0.528  160.387     156631.0            Asaki Takaya    487404.0   
16582    0.558  206.135     135146.0            BluefaceVEVO   6048777.0   
18029    0.551  180.274     174192.0  A Boogie Wit da Hoodie  98998997.0   
19968    0.964  125.308      58872.0               BANGTANTV    954081.0   

          Likes  Comments       Stream  
469       256

Посчитаем среднее значение для комментариев на Youtube для всех значений в датафрейме и для тех, у которых Speechiness > 0.66

In [564]:
mean_comm = round(df['Comments'].mean())
mean_comm_066 = round(speech_066['Comments'].mean())
print ('Cреднее количество «комментариев» на YouTube равно', mean_comm)
print ('Cреднее количество «комментариев» на YouTube для песен с показателем «speechiness» выше 0,66  равно', mean_comm_066)

Cреднее количество «комментариев» на YouTube равно 27519
Cреднее количество «комментариев» на YouTube для песен с показателем «speechiness» выше 0,66  равно 335


Видим, что среднее количество комментариев для всех песен в разы больше, чем для песен со Speechiness > 0.66.
Следовательно, для песен, которые содержат больше слов, оставляют меньше комментариев на Youtube.